dev/benchmarks/test_apps/stocks/lib/main.dart

```diff
@@ -91,19 +89,121 @@ class StocksAppState extends State<StocksApp> {
+    initialRouteName = initialRouteName.substring(1); // strip leading '/'
+    result.add('/');
+    final List<String> routeParts = initialRouteName.split('/');
+    if (initialRouteName.isNotEmpty) {
+      String routeName = '';
+      for (final String part in routeParts) {
+        routeName += '/$part';
+        result.add(routeName);
+      }
+    }
+    return SynchronousFuture<List<String>>(result);
+  }
+
+  @override
+  Future<RoutePackage> restore(List<String> configuration) {
+    return SynchronousFuture<RoutePackage>(
+      RoutePackage(
+        routeName: configuration.last,
+      )
+    );
+  }
+}
+
+class StockRouterDelegate extends RouterDelegate<List<String>> with PopNavigatorRouterDelegateMixin<List<String>> {
+  StockRouterDelegate(this.stockState) {
+    stockState.addListener(notifyListeners);
+  }
+  StocksAppState stockState;
+
+  final HeroController _heroController = MaterialApp.createMaterialHeroController();
+
+  @override
+  final GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();
+
+  @override
+  Future<void> setNewRoutePath(List<String> configuration) {
+    stockState.routesUpdater(configuration);
+    return SynchronousFuture<void>(null);
+  }
+
+  bool _handlePopPage(Route<dynamic> route, dynamic result) {
+    final bool success = route.didPop(result);
+    if (success)
+      stockState.routesUpdater(stockState.routes..removeLast());
+    return success;
+  }
+
+  List<Page<void>> _buildPages() {
+    return List<Page<void>>.from(stockState.routes.map<Page<void>>(_createPage));
+  }
+
+  Page<void> _createPage(String routeName) {
+    switch(routeName) {
+      case '/':
+        return MaterialPage<void>(
+          builder: (BuildContext context) {
+            return StockHome(stockState.stocks, stockState._configuration, stockState.configurationUpdater);
+          }
+        );
+      case '/settings':
+        return MaterialPage<void>(
+          builder: (BuildContext context) {
+            return StockSettings(stockState._configuration, stockState.configurationUpdater);
+          }
+        );
+      default:
+        assert(false); // We should not be asked to create anything else.
+    }
+    return null;
+  }
+
+  Route<dynamic> _getRoute(RouteSettings settings) {
+    if (settings.name == '/stock') {
+      final String symbol = settings.arguments as String;
+      return MaterialPageRoute<void>(
+        settings: settings,
+        builder: (BuildContext context) => StockSymbolPage(symbol: symbol, stocks: stockState.stocks),
+      );
+    }
+    // There should be no other named route.
+    return null;
+  }
+
+
+  @override
+  Widget build(BuildContext context) {
+    return Navigator(
+      key: navigatorKey,
+      pages: _buildPages(),
+      // Enables the material style hero animation.
+      observers: <NavigatorObserver>[_heroController],
+      // We need this call back to support imperative API.
```

Comment:
The idea is to migrate the stock app away from the imperative API to showcase the new declarative API, so this callback should no longer be necessary.

Instead of doing 

https://github.com/flutter/flutter/blob/4b4cff94bf6631bf6326c9239c10b286b4fdb08c/dev/benchmarks/test_apps/stocks/lib/stock_home.dart#L266

The stock app should modify app state and based on the modified app state the RouterDelegate would build a new navigator with a route to show that stock.

Reply:
For that, it may make sense to introduce a new AppState class that's injected into the widget hierarchy via an InheritedWidget. This new AppState class wouldn't inherit from the widget State class. Despite its name, the `StocksAppState` class doesn't really include the app state. It's just the State object for the widget names StockApp.

Reply:
I thought the model we want to present to the developers is that the routerdelegate takes in a  listenable `app state` and builds around it, and we don't emphasis where the state is coming from (through inherited widget or not). I use the  `StocksAppState` in this rewrite because that is considered the main state in the original example.

Also for the imperative api, I was also in the middle of designing the an opinionated router delegate, and was thinking about still giving the api to push named route. I added this for easier migration, do you think we should remove them all together?

I have a WIP repo for this https://github.com/chunhtai/flutter-nested-router/blob/313aadc457c3d661960d4854918deb005b29f65c/lib/router_delegate.dart#L8

with the example on how to use it in https://github.com/chunhtai/flutter-nested-router/blob/master/lib/main.dart